[expotool] generate ah-hoc bare apps via expotool

ajsmth opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

This is the first step for revitalizing `expo-stories` -- in the long run it will hopefully make it easier for the community to install ad-hoc bare apps that have relevant expo modules symlinked:

`et generate-bare-app <appName> [...expoPackages]` 

This will generate an app in the `bare-apps` directory with with whatever `[expoPackages]` were specified, symlinked so that they are easily editable and testable. 

The command also takes a `template` option, which I plan to eventually point to a template that has `expo-stories` initialized, which will simplify generating story apps for users as well, for whichever packages they wish to see stories of. 

# How

<!--
How did you build this feature or fix this bug and why?
-->

The general implementation is: 

- `expo init <template>`
- symlink packages specified to the CLI (and any others that are necessary to get things to work properly)
- `expo prebuild`

Right now there are a few hiccups which is why some of the implementation is a bit strange, and the CLI output could be cleaned up, but iOS and Android builds seem to work in my basic testing - I've added comments in the code to hopefully outline places where this implementation could be improved and cleaned up

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

No test plan as of now, but this command should generate an app that works: 

`et generate-bare-app my-test-app expo-av`

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

tsapeta commented on tools/src/commands/GenerateBareApp.ts:
```diff
@@ -0,0 +1,133 @@
+  const gradleProperties = fs.readFileSync(gradlePropertiesPath, { encoding: 'utf-8' });
+  const updatedGradleProperies = gradleProperties.replace(
+    `FLIPPER_VERSION=0.54.0`,
+    `FLIPPER_VERSION=0.99.0`
```
Could you add a comment why?

ajsmth replied:
do you mean write a comment in the `gradle.properties` file? I added a comment just above this for why I needed to write to this file in case you missed - prebuild is updating it for some reason to an old flipper value

tsapeta replied:
Ah, gotcha! Sorry, I missed that comment 🙈

tsapeta commented on tools/src/commands/GenerateBareApp.ts:
```diff
@@ -0,0 +1,133 @@
+    fs.mkdirSync(projectsDir);
+  }
+
+  if (clean) {
+    if (fs.existsSync(projectDir)) {
+      fs.rmdirSync(projectDir, { recursive: true });
+    }
+  }
```
With `fs.remove` from `fs-extra` you don't have to check if the file exists. See https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/remove.md
```suggestion
  if (clean) {
    await fs.remove(projectDir);
  }
```

ajsmth replied:
🎉 TIL! thank you

tsapeta commented on tools/src/commands/GenerateBareApp.ts:
```diff
@@ -0,0 +1,133 @@
+};
+
+async function action(appName: string, packageNames: string[], options: GenerateBareAppOptions) {
+  function symlinkPackageManually(packageName: string) {
+    const packagePath = path.resolve(projectDir, 'node_modules', packageName);
+    if (fs.existsSync(packagePath)) {
+      fs.rmdirSync(path.resolve(projectDir, 'node_modules', packageName), { recursive: true });
+    }
+
+    fs.symlinkSync(path.resolve(repoRoot, 'packages', packageName), packagePath);
+  }
```
```suggestion
  async function symlinkPackageManually(packageName: string) {
    const packagePath = path.resolve(projectDir, 'node_modules', packageName);
    await fs.ensureSymlink(path.resolve(repoRoot, 'packages', packageName), packagePath);
  }
```

tsapeta commented on tools/src/commands/GenerateBareApp.ts:
```diff
@@ -0,0 +1,133 @@
+
+  await runExpoCliAsync('prebuild', ['--clean'], { cwd: projectDir });
+
+  const ncl = path.resolve(repoRoot, 'apps/native-component-list');
```
I suggest `nclDir` or `nclPath` for clarity 🤔

tsapeta commented on tools/src/commands/GenerateBareApp.ts:
```diff
@@ -0,0 +1,133 @@
+
+  const { clean, outDir = 'bare-apps', template = 'expo-template-bare-minimum' } = options;
+
+  const repoRoot = path.resolve(__dirname, '../../../');
```
This is already defined as `EXPO_DIR` in `tools/src/Constants.ts`

tsapeta commented on tools/src/commands/GenerateBareApp.ts:
```diff
@@ -0,0 +1,133 @@
+import fs from 'fs-extra';
+import { Command } from '@expo/commander';
+import spawnAsync from '@expo/spawn-async';
+import { runExpoCliAsync } from '../ExpoCLI';
```
Doesn't linter require an empty line between dependencies and local imports? 🤔 The dependencies also doesn't seem to be sorted.

```suggestion

import { runExpoCliAsync } from '../ExpoCLI';
```

ajsmth commented on tools/src/commands/GenerateBareApp.ts:
```diff
@@ -0,0 +1,133 @@
+  const pkg = require(path.resolve(projectDir, 'package.json'));
+
+  pkg['expo'] = pkg['expo'] ?? {};
+
+  // TODO - autolinking -> { include?: [] }
+  // pkg['expo']['autolinking'] = { exclude: [] };
+
```
apologies! I was thinking of having an "include" API which would allow a manual override to auto linking -- there are certain modules that are included by default in every project (e.g dev-launcher) that I'd like to explicitly omit because I know I don't need them, but prebuild thinks I do (because monorepo)

I thought that the "exclude" field would achieve this, however the prebuild command uses different logic for determining which modules are present (they do not read from this auto linking field in `package.json`)

ajsmth replied:
I don't think there is a workaround for this at the moment so I might just remove the comment altogether

ajsmth commented:
@tsapeta thanks for all of the feedback - sorry for missing the async / await convention we use - I'll be sure to take a close look next time I'm working on an `expotool` :)

bbarthec reviewed (commented):
Looking good as a starting PR 🤩 
My remarks/questions:
1. Can we add some usage/example to demonstrate full flow of using this command?
2. What about already written ncl screens for every package we have? You plan to move them to separate packages and than include in the generated app? 🤔 
3. Let's split main `action` method into meaningful top-level parts:
```
await cleanPreviousProject();
await runExpoInit();
await modifyPackageJson();
await yarnInstall();
await runExpoPrebuild();
await copyMetroFilesFromNCLProject();
await applyGradleFlipperFixtures();
```
4. Isn't there a more clever way to merge initial package.json's contents with some JSON object? Modifying package.json by `pkg['firstKey']['secondKey']` does not scale very well 😅 Ideally we'll have something like this:
```
const packageJsonModifications = {
  'expo-yarn-wokrspaces': {
    symlinks: [],
  },
  scripts: {
    postinstall: '...'
  },
  ...
}

await mergeJsonModifications(packageJsonPath, packageJsonModifications);
```

ajsmth commented:
@bbarthec I've updated this implementation, most importantly to remove the usage of expo-yarn-workspaces. This lets us generate builds outside of the repo and have them work as expected. Projects built outside of the repo are slimmer because they include a smaller subset of required modules along side the packages specified in the command

Symlinks in the monorepo lead to a bunch of headaches with `prebuild`  applying config plugins for packages that are not part of the target project. That being said, I've still included support for this, it just means builds in the mono repo will include a couple of extra packages

The logic was also updated to figure out relative dependencies to also symlink - for example, `expo-dev-client` relies on ~4 other packages that also need to be symlinked each time, so these are detected via `package.json` instead of having to specify manually each time

Example usage: 

```
1. et generate-bare-app expo-dev-client --name my-app-name --outDir ~/tmp
2. cd ~/tmp/my-app-name
3. expo run:ios
```

ajsmth committed:
[expotool] generate adhoc bare apps

bbarthec reviewed (changes requested):
Looking promising 🤩 
Remarks:
- type everything, I don't see a reason why to skip types for some functions? 🤔 
- consider renaming `projectsDir` to `workspaceDir` as the original name might be easily confused with `projectDir` you're also using in many places

ajsmth commented on tools/src/commands/GenerateBareApp.ts:
```diff
@@ -0,0 +1,237 @@
+  const gradleProperties = await fs.readFile(gradlePropertiesPath, { encoding: 'utf-8' });
+  const updatedGradleProperies = gradleProperties.replace(
+    `FLIPPER_VERSION=0.54.0`,
+    `FLIPPER_VERSION=0.99.0`
```
agreed - this was a patch job just to get things to work

bbarthec replied:
Add a comment maybe and maybe log some info about this? 🤔

ajsmth committed:
Apply suggestions from code review

Co-authored-by: Bartłomiej Bukowski <bartlomiej.bukowski@swmansion.com>

ajsmth committed:
fix typings

ajsmth commented:
thank you @bbarthec ! I've applied your changes an a small typing fix

Next steps will be to make an expo stories template and start actually using this w/ story files, also perhaps testing with more modules and smoothing out some of rough edges

bbarthec reviewed (approved):
Looking good 👍 
Thank you for taking care of the suggestions I've posted.

ajsmth committed:
add log / comment about gradle modification
